src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs

```diff
@@ -346,23 +346,47 @@ internal ImmutableArray<TypeWithAnnotations> SubstituteTypes(ImmutableArray<Type
             }
             else
             {
-                var set = new HashSet<TypeSymbol>();
+                var map = new Dictionary<TypeWithAnnotations, int>(ConstraintEqualsComparer.Instance);
                 foreach (var type in original)
                 {
                     if (ignoreTypesDependentOnTypeParametersOpt == null ||
                         !type.Type.ContainsTypeParameters(ignoreTypesDependentOnTypeParametersOpt))
                     {
                         var substituted = SubstituteType(type);
-                        // TODO: Do we need to merge annotations?
-                        if (set.Add(substituted.Type))
+
+                        if (!map.TryGetValue(substituted, out int mergeWith))
                         {
+                            map.Add(substituted, result.Count);
                             result.Add(substituted);
                         }
+                        else
+                        {
+                            result[mergeWith] = ConstraintsHelper.MergeTopLevelNullabilityForConstraint(result[mergeWith], substituted);
```

Comment:
>MergeTopLevelNullabilityForConstraint [](start = 66, length = 37)

I didn't understand why we need to merge. The `map` uses the strictest type and annotation comparison, so if `map.TryGetValue` found an existing substituted type I would think they are identical already.